In this project I will be working with a dataset from the German eBay website, specifically their car classifieds section. The dataset can be found here: [link](https://data.world/data-society/used-cars-data) Also, I will be working with a modified version of this dataset that has been uncleaned, as the purpose of this project is to work on my cleaning skills. After cleaning the dataset, I will be analyzing the car listings to find notable information about the data. 

In [6]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

Below we can see that there are 20 columns in the dataframe, and 50,000 rows; which agrees with the documention. There is also various information about the listing including the seller, the date the ad was put up, and the fact that various column information is in German including the notRepairedDamage column, the vehicleType column, and the gearbox column.

In [8]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Some other information is that the column names use camelcase, so I will be converting them to snakecase as python works with snakecase better, and it just makes it easier for everyone to read. I will also change some column names to make them easier to understand.

In [10]:
cols = autos.columns



In [64]:
autos.rename({'yearOfRegistration' : 'registration_year'},axis = 1,inplace=True) 
autos.rename({'monthOfRegistration' : 'registration_month'},axis = 1,inplace=True)
autos.rename({'notRepairedDamage' : 'unrepaired_damage'},axis = 1,inplace=True)
autos.rename({'dateCreated' : 'ad_created'},axis = 1,inplace=True)
autos.rename({'nrOfPictures' : 'num_pictures'},axis = 1,inplace=True)
autos.rename({'postalCode' : 'postal_code'},axis = 1,inplace=True)
autos.rename({'lastSeen' : 'last_seen'},axis = 1,inplace=True)
autos.rename({'fuelType' : 'fuel_type'},axis = 1,inplace=True)
autos.rename({'powerPS' : 'power_PS'},axis = 1,inplace=True)
autos.rename({'vehicleType' : 'vehicle_type'},axis = 1,inplace=True)
autos.rename({'offerType' : 'offer_type'},axis = 1,inplace=True)
autos.rename({'abtest' : 'AB_test'},axis = 1,inplace=True)
autos.rename({'dateCrawled' : 'listed_date'},axis = 1,inplace=True)

In [55]:
autos.head()

listed_date                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  AB_test vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [37]:
autos.describe()

price  registration_year      power_PS    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  num_pictures   postal_code  
count        50000.000000       50000.0  50000.000000  
mean             5.723360           0.0  50813.627300  
std              3.711984           0.0  25779.747957  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  30451.000000  
50%              6.000000           0.0  49577.000000  
75%              9.000000           0.0  71540.000000  
max             12.000000           0.0  99998.000000

Here we can see that not all columns that appear to be numeric, are numeric. Price and odometer will need to be cleaned that that we can work with the data later. Also, we can probably drop the num_pictures column as it doesn't appear to show any meaningful data (almost all the values are likely 0). Also, the max values for the registration year and the power in PS are way too high, which is something to make note of later on when working with those columns and I should probablu drop those values. (note: this table was updated after I made these changes below)

In [ ]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos.rename({'odometer' : 'odometer_km'},axis = 1,inplace=True)

In [36]:

autos['odometer_km'] = autos['odometer_km'].astype(float)
autos['price'] = autos['price'].astype(float)

In [56]:
autos.head()


listed_date                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  AB_test vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now I will be looking at the odometer and price columns to find outliers that might be skewing our data.

In [46]:
print(autos['odometer_km'].unique().shape)
print('\n')
print(autos['odometer_km'].describe())
print('\n')
print(autos['odometer_km'].value_counts().sort_index().head())
print('\n')
print(autos['odometer_km'].value_counts().sort_index(ascending = False).head())


(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


5000.0     967
10000.0    264
20000.0    784
30000.0    789
40000.0    819
Name: odometer_km, dtype: int64


150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64


In [52]:
print(autos['price'].unique().shape)
print('\n')
print(autos['price'].describe())
print('\n')
print(autos['price'].value_counts().sort_index().head(10))
print('\n')
print(autos['price'].value_counts().sort_index(ascending = False).head())
autos = autos[autos["price"].between(2,99999999)]

(2354,)


count    4.842200e+04
mean     8.095536e+03
std      1.803248e+05
min      2.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.499000e+03
max      2.732222e+07
Name: price, dtype: float64


2.0     3
3.0     1
5.0     2
8.0     1
9.0     1
10.0    7
11.0    2
12.0    3
13.0    2
14.0    1
Name: price, dtype: int64


27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
Name: price, dtype: int64


The odometer min and max values seemed reasonable, so I left that column alone. However, in the price column we had a value over 99 million dollars so I got rid of that value, and we also had a lot of prices that were one dollar or less (+1000), which does not make sense so I got rid of those values. I wanted to get rid of more outliers, but I was unsure of where to make the distinction of which values seemed plausible or not, so I decided to stop after that as those were the extreme outliers and most likely to skew our data. Now the mean value for cars is about $8,000, which seems pretty reasonable for a used car.

In [81]:
autos['listed_date'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = True)

2016-03-05    0.025360
2016-03-06    0.014064
2016-03-07    0.036037
2016-03-08    0.033311
2016-03-09    0.033063
2016-03-10    0.032196
2016-03-11    0.032588
2016-03-12    0.036967
2016-03-13    0.015654
2016-03-14    0.036616
2016-03-15    0.034261
2016-03-16    0.029511
2016-03-17    0.031639
2016-03-18    0.012907
2016-03-19    0.034736
2016-03-20    0.037793
2016-03-21    0.037338
2016-03-22    0.032940
2016-03-23    0.032258
2016-03-24    0.029387
2016-03-25    0.031556
2016-03-26    0.032237
2016-03-27    0.031122
2016-03-28    0.034840
2016-03-29    0.034158
2016-03-30    0.033724
2016-03-31    0.031824
2016-04-01    0.033724
2016-04-02    0.035480
2016-04-03    0.038578
2016-04-04    0.036512
2016-04-05    0.013073
2016-04-06    0.003160
2016-04-07    0.001384
Name: listed_date, dtype: float64

These values make sense, as the dates are all from 2016, which is what the documentation supports.

In [82]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038825
2016-04-04    0.036884
2016-04-05    0.011792
2016-04-06    0.003242
2016-04-07    0.001239
Name: ad_created, Length: 76, dtype: float64

These values also make sense, as they coincide with the previous values.

In [83]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001074
2016-03-06    0.004337
2016-03-07    0.005411
2016-03-08    0.007393
2016-03-09    0.009644
2016-03-10    0.010615
2016-03-11    0.012370
2016-03-12    0.023811
2016-03-13    0.008860
2016-03-14    0.012618
2016-03-15    0.015861
2016-03-16    0.016439
2016-03-17    0.028066
2016-03-18    0.007331
2016-03-19    0.015819
2016-03-20    0.020631
2016-03-21    0.020631
2016-03-22    0.021375
2016-03-23    0.018587
2016-03-24    0.019743
2016-03-25    0.019185
2016-03-26    0.016811
2016-03-27    0.015613
2016-03-28    0.020900
2016-03-29    0.022345
2016-03-30    0.024741
2016-03-31    0.023853
2016-04-01    0.022862
2016-04-02    0.024865
2016-04-03    0.025195
2016-04-04    0.024493
2016-04-05    0.124943
2016-04-06    0.221593
2016-04-07    0.131985
Name: last_seen, dtype: float64

These values also all seems to fall in the same range as the previous two columns.

In [84]:
autos['registration_year'].describe()

count    48422.000000
mean      2004.772004
std         88.771907
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration year has a mean of 2004, which seems pretty new for a used car selection. However, the max appears to be skewing the results as it has a value of 9999. Also, the min value is 1000, which doesn't make sense as cars were invented in the last century or so. I will drop values less than 1900 since cars were not invented until the early to mid 1900s. Also, we will stop at 2016 since the listing dates were from 2016 at the latest.

In [88]:
autos = autos[autos["registration_year"].between(1900,2016)]
print(autos['registration_year'].value_counts(normalize=True))

2000    0.067208
2005    0.062889
1999    0.062030
2004    0.058012
2003    0.057948
          ...   
1939    0.000021
1948    0.000021
1931    0.000021
1927    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64


Now it makes more sense, the most common cars were registered in the early 2000's, and the least common in the early to mid 1900s. 

In [113]:
brands = autos['brand'].value_counts().head(20).index

I am going to aggregate the top 20 brands of cars in the brand column. This is so I can find the brands with the most data.

In [118]:
top_20_price = {}

for b in brands:
    selected_rows = autos[autos['brand'] == b]
    sorted_rows = selected_rows.sort_values('brand', ascending = False)
    mean_price = sorted_rows.iloc[:,4].mean()
    top_20_price[b] = mean_price

In [119]:
top_20_price

{'volkswagen': 6748.972450950493,
 'bmw': 8606.653117060778,
 'opel': 5460.739843906344,
 'mercedes_benz': 8657.285427807486,
 'audi': 9362.169478908188,
 'ford': 7472.566605279312,
 'renault': 2483.889192886457,
 'peugeot': 3100.6928057553955,
 'fiat': 2818.456066945607,
 'seat': 4412.747058823529,
 'skoda': 6384.667539267016,
 'nissan': 4750.063202247191,
 'mazda': 4112.596614950635,
 'smart': 3580.2239031770046,
 'citroen': 45556.23547400611,
 'toyota': 5167.091062394604,
 'hyundai': 5376.740899357602,
 'sonstige_autos': 46108.511061946905,
 'volvo': 4969.774117647059,
 'mini': 10613.459657701711}

As you can see, some of the most popular brands in Germany are also some the most expensive ones. The notable exceptions are sonstige_autos and mini, which almost seem to be too expensive to be the most popular. Also, the most popular brands are German brands including volkswagen, bmw, and mercedes. These are also fairly expensive brands as noted earlier. Opel is also German, but it is one of the cheaper brands in the top 20.

In [120]:
top_20_mileage = {}

for b in brands:
    selected_rows = autos[autos['brand'] == b]
    sorted_rows = selected_rows.sort_values('brand', ascending = False)
    mean_price = sorted_rows.iloc[:,11].mean()
    top_20_mileage[b] = mean_price
    
top_20_mileage

{'volkswagen': 128715.05540307004,
 'bmw': 132557.16239984366,
 'opel': 129326.59595757455,
 'mercedes_benz': 130838.9037433155,
 'audi': 129208.43672456576,
 'ford': 124188.15224063843,
 'renault': 128144.09484724123,
 'peugeot': 127122.30215827338,
 'fiat': 117066.94560669456,
 'seat': 121029.41176470589,
 'skoda': 110746.07329842931,
 'nissan': 118370.78651685393,
 'mazda': 124464.03385049365,
 'smart': 99326.77760968229,
 'citroen': 119525.99388379205,
 'toyota': 115944.35075885328,
 'hyundai': 106541.75588865097,
 'sonstige_autos': 90000.0,
 'volvo': 138294.11764705883,
 'mini': 88105.13447432763}

In [125]:
top_20_price_series = pd.Series(top_20_price)
top_20_mileage_series = pd.Series(top_20_mileage)
df = pd.DataFrame(top_20_price_series, columns=['mean_price'])
df

mean_price
volkswagen       6748.972451
bmw              8606.653117
opel             5460.739844
mercedes_benz    8657.285428
audi             9362.169479
ford             7472.566605
renault          2483.889193
peugeot          3100.692806
fiat             2818.456067
seat             4412.747059
skoda            6384.667539
nissan           4750.063202
mazda            4112.596615
smart            3580.223903
citroen         45556.235474
toyota           5167.091062
hyundai          5376.740899
sonstige_autos  46108.511062
volvo            4969.774118
mini            10613.459658

In [126]:
df['mean_mileage'] = top_20_mileage_series
df

mean_price   mean_mileage
volkswagen       6748.972451  128715.055403
bmw              8606.653117  132557.162400
opel             5460.739844  129326.595958
mercedes_benz    8657.285428  130838.903743
audi             9362.169479  129208.436725
ford             7472.566605  124188.152241
renault          2483.889193  128144.094847
peugeot          3100.692806  127122.302158
fiat             2818.456067  117066.945607
seat             4412.747059  121029.411765
skoda            6384.667539  110746.073298
nissan           4750.063202  118370.786517
mazda            4112.596615  124464.033850
smart            3580.223903   99326.777610
citroen         45556.235474  119525.993884
toyota           5167.091062  115944.350759
hyundai          5376.740899  106541.755889
sonstige_autos  46108.511062   90000.000000
volvo            4969.774118  138294.117647
mini            10613.459658   88105.134474

Here we can see that the mean mileage tends to be higher with cars that are more popular. All of the top 5 brands have a mean mileage near 130,000, whereas the rest of the brands have vales closer to 120,000. This does make some sense, as people tend to drive the cars that they like more, but it is interesting that cars which are more expensive and have more miles (things people typically associate with being not good) are cars that are the most common on the site. This could also be due to the fact that these cars have reliability problems as they get more miles, so people are preemtively selling their cars (which are worth more) in oder to avoid paying for fees assoicated with these ageing cars. 